feat(editor): Update nodes search to favor full word matches

MiloradFilipovic opened the pull request:
## Summary
Updating our fuzzy search logic to boost matches where the search term appears as a complete word.
![image](https://github.com/user-attachments/assets/b5418d75-1a44-4de7-8ffd-033222b35f46)


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

MiloradFilipovic committed:
feat(editor): Update nodes search to favor full word matches

mutdmour commented on packages/editor-ui/src/utils/sortUtils.ts:
```diff
@@ -240,6 +246,23 @@ export function sublimeSearch<T extends object>(
 				{ value, weight }: { value: string; weight: number },
 			) => {
+				// Check for whole word match first
+				if (containsWord(value, filter)) {
```
can you please add some tests?

mutdmour replied:
have you tried increasing the first word bonus? 
Tried increasing `FIRST_LETTER_BONUS` to 30 and that was enough to get agent to top

mutdmour commented on packages/editor-ui/src/utils/sortUtils.ts:
```diff
@@ -240,6 +246,23 @@ export function sublimeSearch<T extends object>(
+				}
+
+				// Check for substring match (case-insensitive)
+				if (value.toLowerCase().includes(filter.toLowerCase())) {
```
why is this needed too?

mutdmour replied:
I think this might already be addressed by the fuzzy search algorithm

MiloradFilipovic replied:
Indeed, it was a leftover.

mutdmour commented:
testing a bit here. I think this approach needs to be more refined.

search for "code"
before: Code, AI Transform, Cohere Model, Convert to File, extract from file
after: AI Transform, Code, convert to file, extract from file, JWT

much worse afterwards here, especially first result

MiloradFilipovic committed:
Merge branch 'master' into update-fuzzy-search

MiloradFilipovic committed:
👌 Using bonus to bump whole letter matches

MiloradFilipovic committed:
⚡️ Fixing typo in Code node alias

MiloradFilipovic committed:
✅ Adding unit tests

MiloradFilipovic commented:
@mutdmour thanks for testing. It was a bit extreme just to bump scores that much so I introduced another bonus for whole word matches. Take a look if this makes sense (for some reason, just bumping the `FIRST_LETTER_BONUS` didn't work for me).
I also added test for this and fixed a typo we had in Code node alias

MiloradFilipovic committed:
👌 Reverting typo fix

mutdmour commented on packages/nodes-base/nodes/Code/Code.node.json:
```diff
@@ -11,7 +11,7 @@
 			}
 		]
 	},
-	"alias": ["cpde", "Javascript", "JS", "Python", "Script", "Custom Code", "Function"],
```
ya it's not great, but it works. I think the current algorithm should handle it or should be updated to handle it.

MiloradFilipovic committed:
Merge branch 'master' into update-fuzzy-search

MiloradFilipovic committed:
Merge branch 'master' into update-fuzzy-search

MiloradFilipovic committed:
🔥 Removing leftover comment

MiloradFilipovic committed:
✨ Introducing position-based weight calculation

MiloradFilipovic committed:
✅ Adding test for new functionality

MiloradFilipovic committed:
✔️ Updating e2e tests, re-ordering node aliases

CharlieKolb commented:
Superseded by #15302
